### PR TITLE
[AWS] [IAM] Initial Commit

### PR DESCRIPTION
**Pull request by nagypeterjob**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

**What is the current behavior? (You can also link to an open issue here)**

**What is the new behavior (if this is a feature change)?**

**Additional info:**

* [x] add-client-id-to-open-id-connect-provider
* [x] add-role-to-instance-profile
* [x] add-user-to-group
* [x] attach-group-policy
* [x] attach-role-policy
* [x] attach-user-policy
* [x] change-password
* [x] create-access-key
* [x] create-account-alias
* [x] create-group
* [x] create-instance-profile
* [x] create-login-profile
* [x] create-open-id-connect-provider
* [x] create-policy
* [x] create-policy-version
* [x] create-role
* [x] create-saml-provider
* [x] create-service-linked-role
* [x] create-service-specific-credential
* [x] create-user
* [x] create-virtual-mfa-device
* [x] deactivate-mfa-device
* [x] delete-access-key
* [x] delete-account-alias
* [x] delete-account-password-policy
* [x] delete-group
* [x] delete-group-policy
* [x] delete-instance-profile
* [x] delete-login-profile
* [x] delete-open-id-connect-provider
* [x] delete-policy
* [x] delete-policy-version
* [x] delete-role
* [x] delete-role-permissions-boundary
* [x] delete-role-policy
* [x] delete-saml-provider
* [x] delete-server-certificate
* [x] delete-service-linked-role
* [x] delete-service-specific-credential
* [x] delete-signing-certificate
* [x] delete-ssh-public-key
* [x] delete-user
* [x] delete-user-permissions-boundary
* [x] delete-user-policy
* [x] delete-virtual-mfa-device
* [x] detach-group-policy
* [x] detach-role-policy
* [x] detach-user-policy
* [x] enable-mfa-device
* [x] generate-credential-report
* [x] generate-organizations-access-report
* [x] generate-service-last-accessed-details
* [x] get-access-key-last-used
* [x] get-account-authorization-details
* [x] get-account-password-policy
* [x] get-account-summary
* [x] get-context-keys-for-custom-policy
* [x] get-context-keys-for-principal-policy
* [x] get-credential-report
* [x] get-group
* [x] get-group-policy
* [x] get-instance-profile
* [x] get-login-profile
* [x] get-open-id-connect-provider
* [x] get-organizations-access-report
* [x] get-policy
* [x] get-policy-version
* [x] get-role
* [x] get-role-policy
* [x] get-saml-provider
* [x] get-server-certificate
* [x] get-service-last-accessed-details
* [x] get-service-last-accessed-details-with-entities
* [x] get-service-linked-role-deletion-status
* [x] get-ssh-public-key
* [x] get-user
* [x] get-user-policy
* [x] list-access-keys
* [x] list-account-aliases
* [x] list-attached-group-policies
* [x] list-attached-role-policies
* [x] list-attached-user-policies
* [x] list-entities-for-policy
* [x] list-group-policies
* [x] list-groups
* [x] list-groups-for-user
* [x] list-instance-profile-tags
* [x] list-instance-profiles
* [x] list-instance-profiles-for-role
* [x] list-mfa-device-tags
* [x] list-mfa-devices
* [x] list-open-id-connect-provider-tags
* [x] list-open-id-connect-providers
* [x] list-policies
* [x] list-policies-granting-service-access
* [x] list-policy-tags
* [x] list-policy-versions
* [x] list-role-policies
* [x] list-role-tags
* [x] list-roles
* [x] list-saml-provider-tags
* [x] list-saml-providers
* [x] list-server-certificate-tags
* [x] list-server-certificates
* [x] list-service-specific-credentials
* [x] list-signing-certificates
* [x] list-ssh-public-keys
* [x] list-user-policies
* [x] list-user-tags
* [x] list-users
* [x] list-virtual-mfa-devices
* [x] put-group-policy
* [x] put-role-permissions-boundary
* [x] put-role-policy
* [x] put-user-permissions-boundary
* [x] put-user-policy
* [x] remove-client-id-from-open-id-connect-provider
* [x] remove-role-from-instance-profile
* [x] remove-user-from-group
* [x] reset-service-specific-credential
* [x] resync-mfa-device
* [x] set-default-policy-version
* [x] set-security-token-service-preferences
* [x] simulate-custom-policy
* [x] simulate-principal-policy
* [x] tag-instance-profile
* [x] tag-mfa-device
* [x] tag-open-id-connect-provider
* [x] tag-policy
* [x] tag-role
* [x] tag-saml-provider
* [x] tag-server-certificate
* [x] tag-user
* [x] untag-instance-profile
* [x] untag-mfa-device
* [x] untag-open-id-connect-provider
* [x] untag-policy
* [x] untag-role
* [x] untag-saml-provider
* [x] untag-server-certificate
* [x] untag-user
* [x] update-access-key
* [x] update-account-password-policy
* [x] update-assume-role-policy
* [x] update-group
* [x] update-login-profile
* [x] update-open-id-connect-provider-thumbprint
* [x] update-role
* [x] update-role-description
* [x] update-saml-provider
* [x] update-server-certificate
* [x] update-service-specific-credential
* [x] update-signing-certificate
* [x] update-ssh-public-key
* [x] update-user
* [x] upload-server-certificate
* [x] upload-signing-certificate
* [x] upload-ssh-public-key
* [x] wait

### CODE REVIEW

File: dev/aws/iam.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { findConfigFile } from "typescript";
+import { FileTextChanges } from "typescript";
```
Comment (cstrnt) [inline dev/aws/iam.ts]: We don't need this

File: dev/aws/iam.ts [hunk 2]
```diff
@@ -335,21 +341,37 @@ const postPrecessGenerator = (
...
+    for (let i = 0; i < enabled.length; i++) {
+      const out = await executeShellCommand(enabled[i]["command"]);
+      list[i] = postPrecessGenerator(
+        out,
+        enabled[i]["parentKey"],
+        enabled[i]["childKey"]
+      );
+    }
+
+    const merged = Array.prototype.concat.apply([], list) as Fig.Suggestion[];
```
Comment (cstrnt) [inline dev/aws/iam.ts:359]: What does this line do? I think we should either add a comment here or use more human-readable code. I think you might be looking for `.flat()` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat)

File: dev/aws/iam.ts [hunk 3]
```diff
@@ -181,6 +181,12 @@ const awsPrincipals = [
...
 
+const identityStruct = [
+  { command: "aws iam list-users", parentKey: "Users", childKey: "Arn" },
+  { command: "aws iam list-groups", parentKey: "Groups", childKey: "Arn" },
+  { command: "aws iam list-roles", parentKey: "Roles", childKey: "Arn" },
+];
+
 const appendFolderPath = (tokens: string[], prefix: string): string => {
   var baseLSCommand = "\\ls -1ApL ";
   var whatHasUserTyped = tokens[tokens.length - 1];
```
Comment (cstrnt) [inline dev/aws/iam.ts]: Could you please use `const` over `var`

**Review by cstrnt — CHANGES_REQUESTED** [review]
I never used AWS therefore this is just a code-side review from me :)

File: dev/aws/iam.ts (hunk 3 above)
Comment (nagypeterjob) [reply dev/aws/iam.ts]: done

File: dev/aws/iam.ts (hunk 2 above)
Comment (nagypeterjob) [reply dev/aws/iam.ts]: thanks, fixed

File: dev/aws/iam.ts (hunk 1 above)
Comment (nagypeterjob) [reply dev/aws/iam.ts]: hmm you have checked some outdated commit :D

File: dev/aws/iam.ts [hunk 4]
```diff
@@ -1,3 +1,958 @@
...
+
+const appendFolderPath = (tokens: string[], prefix: string): string => {
+  const baseLSCommand = "\\ls -1ApL ";
+  let whatHasUserTyped = tokens[tokens.length - 1];
+
+  if (whatHasUserTyped.startsWith(prefix)) {
+    whatHasUserTyped = whatHasUserTyped.slice(prefix.length);
+  } else {
+    return `echo '${prefix}'`;
+  }
```
Comment (cstrnt) [inline dev/aws/iam.ts:198]: Just a tiny change, but reverting the if allows us to omit  the else block and make the code more readable imho
```suggestion
  if (!whatHasUserTyped.startsWith(prefix)) {
     return `echo '${prefix}'`;
  } 
  whatHasUserTyped = whatHasUserTyped.slice(prefix.length);
```

File: dev/aws/iam.ts [hunk 5]
```diff
@@ -1,3 +1,958 @@
...
+    return [
+      ...other_arr.sort(sortFnStrings),
+      "../",
+      ...dots_arr.sort(sortFnStrings),
+    ];
+  };
+
+  const temp_array = alphabeticalSortFilesAndFolders(out.split("\n"));
+
+  const final_array = [];
```
Comment (cstrnt) [inline dev/aws/iam.ts:247]: I think @QuiiBz mentioned it once: We really don't have any variable naming conventions, but in the JS/TS World we mostly use camelCase for variables

File: dev/aws/iam.ts [hunk 6]
```diff
@@ -1,3 +1,958 @@
...
+      if (childKey) {
+        return {
+          name: elm[childKey],
+        };
+      }
+
+      return {
+        name: elm,
+      };
+    });
```
Comment (cstrnt) [inline dev/aws/iam.ts:312]:
```ts
    return policies.map((elm) => childKey ? elm[childKey] : elm)
```

Comment (cstrnt) [reply dev/aws/iam.ts:312]: possible readability improvement

File: dev/aws/iam.ts [hunk 7]
```diff
@@ -1,3 +1,958 @@
...
+      if (childKey) {
+        return {
+          name: item[childKey],
+        };
+      }
+
+      return {
+        name: item,
+      };
+    });
```
Comment (cstrnt) [inline dev/aws/iam.ts:336]: see above

File: dev/aws/iam.ts [hunk 8]
```diff
@@ -1,3 +1,958 @@
...
+    for (let i = 0; i < enabled.length; i++) {
+      const out = await executeShellCommand(enabled[i]["command"]);
+      list[i] = postPrecessGenerator(
+        out,
+        enabled[i]["parentKey"],
+        enabled[i]["childKey"]
+      );
+    }
+
+    return list.flat();
```
Comment (cstrnt) [inline dev/aws/iam.ts:359]: Does this need to run sequentially? If not we could use `Promise.all` to run them in parallel which could be a performance gain

File: dev/aws/iam.ts [hunk 9]
```diff
@@ -1,3 +1,958 @@
...
+            name: elm["RoleName"],
+          };
+        });
+      } catch (e) {
+        console.log(e);
+      }
+      return [];
+    },
+    cache: {
+      ttl: 30000,
```
Comment (cstrnt) [inline dev/aws/iam.ts:461]: Same as in the other Spec. I think we should create a const for the cache TTL

**Review by cstrnt — CHANGES_REQUESTED** [review]

**Comment by QuiiBz** [comment]
Hello, is this PR still in WIP?

**Comment by nagypeterjob** [comment]
> Hello, is this PR still in WIP?

Not anymore, I just cannot assign / modify labels

**Comment by QuiiBz** [comment]
> 
> Not anymore, I just cannot assign / modify labels

Alright thanks, apologize for the delay on this PR. We're going to merge this asap!

**Review by QuiiBz — APPROVED** [review]
Such a huge PR, thank you so much! This is going to be really helpful

**Pull request merged**